docs: add docs about list-like feature (#762)
docarray/docarray: docs/advanced/document-store/extend.md

Note that DocumentArray maintains an `offset2ids` mapping to allow a list-like behaviour. This mapping is 
inherited from the `BaseGetSetDelMixin`. Therefore, you need to implement methods to persist this mapping, in case you 
want to also persist the ordering of Documents inside the storage. Considered the performance issue brought by list-like
structure implemented by Offset2id, you can disable this feature by passing a flag when constructing the backend.
After so, the Offset2id will become a shell and do nothing.

```Python
from docarray import DocumentArray, Document

> da = DocumentArray(storage='qdrant', config={'n_dim':3, 'list_like':False})

Author: Make sense

Note that DocumentArray maintains an `offset2ids` mapping to allow a list-like behaviour. This mapping is 
inherited from the `BaseGetSetDelMixin`. Therefore, you need to implement methods to persist this mapping, in case you 

> want to also persist the ordering of Documents inside the storage. Considered the performance cost brought by list-like
> structure implemented by Offset2id, you can disable this feature by passing a flag when constructing the backend.
> After so, the Offset2id will become a shell and do nothing. Then in the `getsetdel.py` you can construct the _offset2ids 
> member variable by passing list_like flag as follows:

Reviewer: ```suggestion
want to also persist the ordering of Documents inside the storage. However, the list-like
structure implemented by Offset2id can introduce performance bottlenecks, so you can disable this feature by passing a flag when constructing the backend.

In your backend implementations, in `getsetdel.py`, you have to construct the _offset2ids 
member variable by passing list_like flag as follows:
```

> want to also persist the ordering of Documents inside the storage. Considered the performance cost brought by list-like
> structure implemented by Offset2id, you can disable this feature by passing a flag when constructing the backend.
> After so, the Offset2id will become a shell and do nothing. Then in the `getsetdel.py` you can construct the _offset2ids 
> member variable by passing list_like flag as follows:

Reviewer 2: ```suggestion
want to also persist the ordering of Documents inside the storage. Considering the performance cost brought by the list-like structure implemented by Offset2id, you can disable this feature by passing a flag when constructing the backend.
After this, the Offset2id will become a shell and do nothing. Then in the `getsetdel.py` you can construct the `_offset2ids` 
member variable by passing the `list_like` flag as follows:
```

Note that DocumentArray maintains an `offset2ids` mapping to allow a list-like behaviour. This mapping is 
inherited from the `BaseGetSetDelMixin`. Therefore, you need to implement methods to persist this mapping, in case you 
want to also persist the ordering of Documents inside the storage. However, the list-like

> structure implemented by Offset2id can introduce performance bottlenecks, so you can disable this feature by passing a flag when constructing the backend.

Reviewer: ```suggestion
structure implemented by Offset2id can introduce performance bottlenecks, so you can disable this feature by passing a flag when constructing the backend.

In step 4 you will see how you should read in this flag (`list_like`) from the user. Here you have to use it do adapt some operations.
```

As a reference, to see how we implement for SQLite, check out {class}`~docarray.array.storage.sqlite.seqlike.SequenceLikeMixin`.
```

> For the reason to support the list-like feature, the list-like APIs should do flag checking only when the offset2id structure is called as follows:

Reviewer 2: ```suggestion
To support the list-like feature, the list-like APIs should perform flag checking only when the offset2id structure is called as follows:
```

`MyDocStoreConfig` is a dataclass for containing the configs. You can expose arguments of your document store to this data class and allow users to customize them. In `init_storage` function, you need to parse `config` either from `MyDocStoreConfig` object or a `Dict`.

> The `list-like` flag is passed into the config file as follows:

Reviewer: ```suggestion
To enable the disabling of list-like features, your configuration should accept the flag `list_like` as follows:
```

```python
@dataclass
class MyDocStoreConfig:
    config1: str
    config2: str

>     list - like: True

Reviewer 3: ```suggestion
    list_like: bool
```

    config3: Dict
    ...
```

> By default, it is set to be `True`. 

Reviewer 2: ```suggestion
By default, it is set to `True`. 
```

> By default, it is set to be `True`. 

Reviewer: ```suggestion
By default, this should be set to be `True`.

Further, you have to store the value of this flag in `self._list_like`. Some methods that are handled outside of your control will take the value form there and use it appropriately.
```
